### PR TITLE
[String] allow passing a string of custom characters to ByteString::fromRandom

### PR DESCRIPTION
**Pull request by azjezz**
| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | no
| New feature?  | yes/
| Deprecations? | no 
| License       | MIT
| Doc PR        | symfony/symfony-docs#...

### CODE REVIEW

File: src/Symfony/Component/String/ByteString.php [hunk 1]
```diff
@@ -25,20 +25,58 @@
...
 
-    public static function fromRandom(int $length = 16): self
+    public static function fromRandom(int $length = 16, string $alphabet = null): self
     {
-        $string = '';
+        if (0 === $length) {
+            return new static('');
+        }
+
+        if ($length < 0) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/String/ByteString.php:41]: let's turn this into an `<=` and remove the previous if

File: src/Symfony/Component/String/ByteString.php [hunk 2]
```diff
@@ -25,20 +25,58 @@
...
+            throw new \RuntimeException(sprintf('Expected positive length value, got "%d".', $length));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/String/ByteString.php:42]: InvalidArgumentException in the namespace of the component

File: src/Symfony/Component/String/ByteString.php [hunk 3]
```diff
@@ -25,20 +25,58 @@
...
+        }
+
+        $alphabet = $alphabet ?? self::ALPHABET_ALPHANUMERIC;
+        $alphabet_size = \strlen($alphabet);
+        $bits = (int) ceil(log($alphabet_size, 2.0));
+        if ($bits <= 0 || $bits > 56) {
+            throw new \RuntimeException('Expected $alphabet\'s length to be in [2^1, 2^56].');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/String/ByteString.php:49]: InvalidArgumentException

File: src/Symfony/Component/String/ByteString.php [hunk 4]
```diff
@@ -25,20 +25,58 @@
  */
 class ByteString extends AbstractString
 {
+    private const ALPHABET_ALPHANUMERIC = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/String/ByteString.php:28]: there are only 62 chars here, the previous had 64.
can you bench the previous implementation with the new one in the base64 case please?
we might decide to keep a fast path for base64, if it's faster.

File: src/Symfony/Component/String/ByteString.php [hunk 5]
```diff
@@ -25,20 +25,58 @@
...
+        if (0 === $length) {
+            return new static('');
+        }
+
+        if ($length < 0) {
+            throw new \RuntimeException(sprintf('Expected positive length value, got "%d".', $length));
+        }
+
+        $alphabet = $alphabet ?? self::ALPHABET_ALPHANUMERIC;
+        $alphabet_size = \strlen($alphabet);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/String/ByteString.php:46]: camel case everywhere

File: src/Symfony/Component/String/ByteString.php [hunk 6]
```diff
@@ -25,20 +25,58 @@
...
+            $data = random_bytes($urandom_length);
+            $unpacked_data = 0;
+            $unpacked_bits = 0;
+            for ($i = 0; $i < $urandom_length && $length > 0; ++$i) {
+                // Unpack 8 bits
+                $unpacked_data = ($unpacked_data << 8) | \ord($data[$i]);
+                $unpacked_bits += 8;
+
+                // While we have enough bits to select a character from the alphabet, keep
+                // consuming the random data
```
Comment (nicolas-grekas) [inline src/Symfony/Component/String/ByteString.php:64]: we need to give credit to the original implementation - it's MIT btw if anyone is wondering:
https://github.com/hhvm/hsl/blob/master/src/random/private.php

File: src/Symfony/Component/String/CHANGELOG.md [hunk 7]
```diff
@@ -12,6 +12,7 @@ CHANGELOG
    depending of the input string UTF-8 compliancy
  * added `$cut` parameter to `Symfony\Component\String\AbstractString::truncate()`
  * added `AbstractString::containsAny()`
+ * allow passing a string of custom characters to ByteString::fromRandom
```
Comment (nicolas-grekas) [inline src/Symfony/Component/String/CHANGELOG.md:15]: with trailing brackets and backticks

File: src/Symfony/Component/String/Tests/ByteStringTest.php [hunk 8]
```diff
@@ -12,6 +12,7 @@
 namespace Symfony\Component\String\Tests;
 
 use Symfony\Component\String\AbstractString;
+use function Symfony\Component\String\b;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/String/Tests/ByteStringTest.php:15]: not sure about the order, it's strange to have it in the middle

File: src/Symfony/Component/String/ByteString.php [hunk 9]
```diff
@@ -25,20 +25,58 @@
...
+                for (; $unpacked_bits >= $bits && $length > 0; $unpacked_bits -= $bits) {
+                    $index = ($unpacked_data & ((1 << $bits) - 1));
+                    $unpacked_data >>= $bits;
+                    // Unfortunately, the alphabet size is not necessarily a power of two.
+                    // Worst case, it is 2^k + 1, which means we need (k+1) bits and we
+                    // have around a 50% chance of missing as k gets larger
+                    if ($index < $alphabet_size) {
+                        $ret .= $alphabet[$index];
```
Comment (nicolas-grekas) [inline src/Symfony/Component/String/ByteString.php:72]: funny idea: move the function on the base class and allow "characters" to be either code points and/or grapheme clusters, depending on what chunk() returns

**Review by nicolas-grekas — COMMENTED** [review]
Thanks, here are some early random comments :)

File: src/Symfony/Component/String/ByteString.php (hunk 4 above)
Comment (azjezz) [reply src/Symfony/Component/String/ByteString.php:28]: the previous had 64 but `+` and `/` were always being replaced by `''`, so it is way always 62

Comment (nicolas-grekas) [reply src/Symfony/Component/String/ByteString.php:28]: right I forgot about this :)
what about the perf?

File: src/Symfony/Component/String/ByteString.php (hunk 1 above)
Comment (azjezz) [reply src/Symfony/Component/String/ByteString.php:41]: that would mean `ByteString::fromRandom(0)` results in exception? :thinking:

Comment (nicolas-grekas) [reply src/Symfony/Component/String/ByteString.php:51]: Yes :)

File: src/Symfony/Component/String/ByteString.php (hunk 4 above)
Comment (azjezz) [reply src/Symfony/Component/String/ByteString.php:28]: not sure tbh, but I would assume the previous implmenetation have a better performance when it comes to using the alphanumeric ( no custome characters )

File: src/Symfony/Component/String/Tests/ByteStringTest.php (hunk 8 above)
Comment (azjezz) [reply src/Symfony/Component/String/Tests/ByteStringTest.php:15]: I suppose Symfony style doesn't have a rule about functions, here, the alphabet order is used.

I think the Symfony coding style can be updated to have function separated (before or after other types )?

but that's out of the scope of the PR :) this is simply following Symfony coding style.

**Review by nicolas-grekas — APPROVED** [review]
let's go and improve during the betas if there is something to improve

**Comment by fabpot** [comment]
Thank you @azjezz.

**Pull request merged**